GitHub repo: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>
Commit ID: aeb2b8fdb97650df4f06d523cf6492a4d25b3e5b

In [160]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [94]:
# Load the dataset (load remotely or locally)
data = pd.read_csv('iris.csv')

# Output the first 15 rows of the data
print(data.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(data.info())
print(data.describe())


    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The features of the data are the Sepal length and width and the Petal length and width. The labels are the three species of flower given: virginica, versicolor and setosa. we map setosa to 0, versicolor to 1 and virginica to 2

# Part 2: Split the dataset into train and test

In [151]:
# Take the dataset and split it into our features (X) and label (y)
X = data.drop(columns=['species'])
y = data['species']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=16)



# Part 3: Logistic Regression

In [152]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression(max_iter=200)
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data = pd.DataFrame([[5.5, 2.3, 4.0, 1.3]], columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])
probabilities = model.predict_proba(sample_data)
print(probabilities)

# iii. Report on the score for Logistic regression model, what does the score measure?
score = model.score(X_test, y_test)
print(score)

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = model.coef_
intercepts = model.intercept_
print(coefficients, intercepts)

[[0.01666176 0.94881898 0.03451926]]
0.9
[[-0.45490623  0.81554438 -2.33586898 -1.00043355]
 [ 0.15641431 -0.50901935 -0.00908929 -0.96012434]
 [ 0.29849192 -0.30652503  2.34495828  1.9605579 ]] [  9.68010393   4.0469188  -13.72702274]


The score reflects the proportion of correct predictions made by the model, An accuracy of 0.9 indicates a solid performance, although it may be limited by its linear assumptions, potentially overlooking more intricate relationships.

# Part 4: Support Vector Machine

In [153]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svm_model = SVC(probability=True)
svm_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data = pd.DataFrame([[5.5, 2.3, 4.0, 1.3]], columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])
probabilities = svm_model.predict_proba(sample_data)
print(probabilities)

# iii. Report on the score for the SVM, what does the score measure
score = svm_model.score(X_test, y_test)
print(score)

[[0.0076329 0.9845263 0.0078408]]
0.9333333333333333


The score reflects the proportion of correct predictions made by the model, An accuracy of 0.933 suggests that the model successfully differentiates between classes, leveraging its strength in handling complex boundaries.

# Part 5: Neural Network

In [158]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(hidden_layer_sizes=(10, 10), max_iter=2000, random_state=42)
mlp_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data = pd.DataFrame([[5.5, 2.3, 4.0, 1.3]], columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])
probabilities = mlp_model.predict_proba(sample_data)
print(probabilities)

# iii. Report on the score for the Neural Network, what does the score measure?
score = mlp_model.score(X_test, y_test)
print(score)

# iv: Experiment with different options for the neural network, report on your best configuration
best_score = 0
best_config = None

configurations = [
    {'hidden_layer_sizes': (10, 10), 'activation': 'relu', 'solver': 'adam'},
    {'hidden_layer_sizes': (15, 15), 'activation': 'tanh', 'solver': 'adam'},
    {'hidden_layer_sizes': (5, 5, 5), 'activation': 'relu', 'solver': 'sgd'},
    {'hidden_layer_sizes': (20,), 'activation': 'logistic', 'solver': 'adam'},
    {'hidden_layer_sizes': (10, 5), 'activation': 'tanh', 'solver': 'lbfgs'},
]

for config in configurations:
    mlp = MLPClassifier(hidden_layer_sizes=config['hidden_layer_sizes'], activation=config['activation'],
                        solver=config['solver'], max_iter=2000, random_state=42)
    mlp.fit(X_train, y_train)
    config_score = mlp.score(X_test, y_test)
    if config_score > best_score:
        best_score = config_score
        best_config = config

print("Best configuration:", best_config, ", accuracy:", best_score)

[[0.00254237 0.82405877 0.17339886]]
0.9666666666666667
Best configuration: {'hidden_layer_sizes': (15, 15), 'activation': 'tanh', 'solver': 'adam'} , accuracy: 1.0


The score reflects the proportion of correct predictions made by the model, An accuracy of 0.97 indicates that the model effectively captures the intricacies of the data, suggesting a high level of reliability in its predictions.

# Part 6: K-Nearest Neighbors

In [174]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
k = 100
knn_model = KNeighborsClassifier(k)
knn_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data = pd.DataFrame([[5.5, 2.3, 4.0, 1.3]], columns=['sepal_length', 'sepal_width', 'petal_length', 'petal_width'])
probabilities = knn_model.predict_proba(sample_data)
print(probabilities)

# iii. Report on the score for kNN, what does the score measure?
score = knn_model.score(X_test, y_test)
print(score)

[[0.26 0.38 0.36]]
0.7


The score reflects the proportion of correct predictions made by the model, An accuracy of 0.7 reveals that the model struggles with classification, possibly due to its sensitivity to parameter settings and data characteristics, indicating room for improvement.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The results of this notebook show how different classification models can significantly change the accuracy of a prediction made. The highest-performing model was the Neural Network achieving a score of 0.97. This might be because of the model's multilayer structure, which allows it to capture complex patterns. What surprised me was the variation in the score of each model, especially the K-nearest neighbours model. It only achieved a score of 0.7, this could have changed if I moved around the value of k.